core/src/test/java/org/apache/iceberg/data/avro/TestDecoderResolver.java

```diff
@@ -0,0 +1,155 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.data.avro;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.apache.avro.Schema;
+import org.apache.avro.io.Decoder;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.ResolvingDecoder;
+import org.apache.iceberg.ManifestFile;
+import org.apache.iceberg.avro.AvroSchemaUtil;
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestDecoderResolver {
+
+  @Before
+  public void before() {
+    DecoderResolver.DECODER_CACHES.get().clear();
+  }
+
+  @Test
+  public void testDecoderCachingReadSchemaSameAsFileSchema() throws Exception {
+    Decoder dummyDecoder = DecoderFactory.get().binaryDecoder(new byte[] {}, null);
+    Schema fileSchema = avroSchema();
+    ResolvingDecoder resolvingDecoder =
+        DecoderResolver.resolve(dummyDecoder, fileSchema, fileSchema);
+
+    assertThat(DecoderResolver.DECODER_CACHES.get().size()).isEqualTo(1);
+    assertThat(DecoderResolver.DECODER_CACHES.get().get(fileSchema).size()).isEqualTo(1);
+    checkCached(fileSchema, fileSchema);
+
+    // Equal but new one
+    Schema fileSchema1 = avroSchema();
+    assertThat(fileSchema1).isEqualTo(fileSchema);
+    ResolvingDecoder resolvingDecoder1 =
+        DecoderResolver.resolve(dummyDecoder, fileSchema1, fileSchema1);
+    assertThat(resolvingDecoder1).isSameAs(resolvingDecoder);
+
+    assertThat(DecoderResolver.DECODER_CACHES.get().size()).isEqualTo(1);
+    assertThat(DecoderResolver.DECODER_CACHES.get().get(fileSchema1).size()).isEqualTo(1);
+    checkCached(fileSchema1, fileSchema1);
+
+    // New one
+    Schema fileSchema2 = avroSchema("manifest_path", "manifest_length");
+    ResolvingDecoder resolvingDecoder2 =
+        DecoderResolver.resolve(dummyDecoder, fileSchema2, fileSchema2);
+    assertThat(resolvingDecoder2).isNotSameAs(resolvingDecoder);
+
+    assertThat(DecoderResolver.DECODER_CACHES.get().size()).isEqualTo(2);
+    assertThat(DecoderResolver.DECODER_CACHES.get().get(fileSchema2).size()).isEqualTo(1);
+    checkCached(fileSchema2, fileSchema2);
+
+    fileSchema = null;
+    System.gc();
+    // Wait the weak reference keys are GCed
+    Thread.sleep(1000);
```

Comment:
nit: Do we want to use Awaitility, like mentioned here: #7773?

Reply:
Thanks @pvary for reviewing. Replaced with Awaitility.